File: src/utils/imageQuality.ts
import { QualityCheckResult, QualityMetrics } from "../types";

/**
 * Analyzes leaf photograph using HTML5 Canvas to measure:
 * 1. Image resolution & aspect ratio
 * 2. Mean luminosity (detects severe underexposure or blinding glare)
 * 3. Laplacian Variance approximation (detects camera defocus / motion blur)
 */
export async function analyzeImageQuality(imageSource: string): Promise<QualityCheckResult> {
  return new Promise((resolve) => {
    const img = new Image();
    img.crossOrigin = "anonymous";
    img.src = imageSource;

    img.onload = () => {
      const originalWidth = img.naturalWidth || img.width;
      const originalHeight = img.naturalHeight || img.height;
      const aspectRatio = Number((originalWidth / (originalHeight || 1)).toFixed(2));

      // Use a standardized working canvas (max 320px) for consistent, high-performance edge & luminance analysis
      const analysisWidth = Math.min(originalWidth, 300);
      const analysisHeight = Math.min(originalHeight, 300);

      const canvas = document.createElement("canvas");
      canvas.width = analysisWidth;
      canvas.height = analysisHeight;
      const ctx = canvas.getContext("2d");

      if (!ctx) {
        resolve(createFallbackResult(originalWidth, originalHeight));
        return;
      }

      ctx.drawImage(img, 0, 0, analysisWidth, analysisHeight);
      const imageData = ctx.getImageData(0, 0, analysisWidth, analysisHeight);
      const data = imageData.data;
      const totalPixels = analysisWidth * analysisHeight;

      let sumLuminosity = 0;
      const gray = new Float32Array(totalPixels);

      for (let i = 0; i < totalPixels; i++) {
        const r = data[i * 4];
        const g = data[i * 4 + 1];
        const b = data[i * 4 + 2];
        // Standard perceived luminance
        const lum = 0.299 * r + 0.587 * g + 0.114 * b;
        sumLuminosity += lum;
        gray[i] = lum;
      }

      const meanLuminosity = Math.round(sumLuminosity / totalPixels);

      // Discrete Laplacian 3x3 kernel:
      // [ 0,  1,  0]
      // [ 1, -4,  1]
      // [ 0,  1,  0]
      let laplacianSum = 0;
      let laplacianSqSum = 0;
      let count = 0;

      for (let y = 1; y < analysisHeight - 1; y++) {
        for (let x = 1; x < analysisWidth - 1; x++) {
          const idx = y * analysisWidth + x;
          const up = (y - 1) * analysisWidth + x;
          const down = (y + 1) * analysisWidth + x;
          const left = y * analysisWidth + (x - 1);
          const right = y * analysisWidth + (x + 1);

          const lap = gray[up] + gray[down] + gray[left] + gray[right] - 4 * gray[idx];
          laplacianSum += lap;
          laplacianSqSum += lap * lap;
          count++;
        }
      }

      const meanLap = laplacianSum / (count || 1);
      const varianceLap = Math.max(0, laplacianSqSum / (count || 1) - meanLap * meanLap);
      const blurScore = Math.round(varianceLap * 10) / 10;

      const issues: string[] = [];
      const recommendations: string[] = [];

      // 1. Resolution Check (MobileNetV3 requires 224x224 minimum)
      if (originalWidth < 224 || originalHeight < 224) {
        issues.push("Low Resolution: Image is below the recommended 224x224 pixel dimension.");
        recommendations.push("Move closer to the crop leaf or use higher resolution photo mode.");
      }

      // 2. Luminosity & Exposure Check
      if (meanLuminosity < 38) {
        issues.push("Underexposed / Too Dark: Leaf symptoms and fungal lesions cannot be clearly differentiated in shadows.");
        recommendations.push("Move leaf into indirect natural daylight or turn on flashlight/torch.");
      } else if (meanLuminosity > 228) {
        issues.push("Overexposed / High Glare: Intense light reflections wash out foliar details.");
        recommendations.push("Angle the camera to shield direct midday sunlight glare off the leaf surface.");
      }

      // 3. Focus & Blur Check
      if (blurScore < 45) {
        issues.push("Blur Detected: The image lacks sharp contrast transitions along leaf veins and lesion margins.");
        recommendations.push("Keep hands steady, tap screen to focus specifically on the leaf spot, and retake.");
      }

      const isValid = issues.length === 0;

      if (!isValid && recommendations.length === 0) {
        recommendations.push("Position a single leaf flat against a clean neutral background.");
      }

      const metrics: QualityMetrics = {
        width: originalWidth,
        height: originalHeight,
        aspectRatio,
        meanLuminosity,
        blurScore
      };

      resolve({
        isValid,
        issues,
        recommendations,
        metrics
      });
    };

    img.onerror = () => {
      resolve({
        isValid: false,
        issues: ["Unable to load or decode image data."],
        recommendations: ["Ensure the file is a valid JPG, PNG, or WebP image."],
        metrics: {
          width: 0,
          height: 0,
          aspectRatio: 1,
          meanLuminosity: 0,
          blurScore: 0
        }
      });
    };
  });
}

function createFallbackResult(w: number, h: number): QualityCheckResult {
  return {
    isValid: true,
    issues: [],
    recommendations: ["Ensure leaf is clearly lit and in focus."],
    metrics: {
      width: w,
      height: h,
      aspectRatio: Number((w / (h || 1)).toFixed(2)),
      meanLuminosity: 128,
      blurScore: 120
    }
  };
}

File: core/image_quality.py
"""
AgroScan AI — Image Quality Assurance Pipeline
Pre-flight check before running neural network screening:
- Valid format and readable image
- Minimum resolution (>= 224x224)
- Blur detection via Laplacian variance
- Underexposure (extreme darkness) and Overexposure (harsh glare) detection
- Actionable farmer guidance
"""

import os
from pathlib import Path
from typing import Dict, Any, List

try:
    from PIL import Image, ImageStat
    import numpy as np
except ImportError:
    Image = None
    ImageStat = None
    np = None


class QualityCheckResult:
    def __init__(self, is_valid: bool, issues: List[str], recommendations: List[str], metrics: Dict[str, Any]):
        self.is_valid = is_valid
        self.issues = issues
        self.recommendations = recommendations
        self.metrics = metrics

    def to_dict(self) -> Dict[str, Any]:
        return {
            "is_valid": self.is_valid,
            "issues": self.issues,
            "recommendations": self.recommendations,
            "metrics": self.metrics
        }


def check_image_quality(image_path: str, min_size: int = 224, blur_threshold: float = 80.0) -> QualityCheckResult:
    """
    Evaluates whether an uploaded or captured crop leaf image is suitable for AI analysis.
    """
    issues = []
    recommendations = []
    metrics = {
        "width": 0,
        "height": 0,
        "mean_luminosity": 0.0,
        "blur_score": 0.0,
        "aspect_ratio": 1.0
    }

    path = Path(image_path)
    if not path.exists():
        return QualityCheckResult(
            is_valid=False,
            issues=["Image file does not exist on disk."],
            recommendations=["Please select an existing leaf photograph."],
            metrics=metrics
        )

    if Image is None:
        # Fallback if PIL is not installed
        return QualityCheckResult(is_valid=True, issues=[], recommendations=[], metrics=metrics)

    try:
        with Image.open(path) as img:
            # Check format
            fmt = (img.format or "").upper()
            if fmt not in ["JPEG", "JPG", "PNG", "WEBP", "BMP"]:
                issues.append(f"Unsupported image format ({fmt}). Please use JPG or PNG.")

            w, h = img.size
            metrics["width"] = w
            metrics["height"] = h
            metrics["aspect_ratio"] = round(w / max(1, h), 2)

            # Minimum resolution check
            if w < min_size or h < min_size:
                issues.append(f"Image resolution too low ({w}x{h} px). Minimum required is {min_size}x{min_size} px.")
                recommendations.append("Move closer to the leaf or take a higher resolution photo.")

            # Convert to RGB for lighting analysis
            rgb_img = img.convert("RGB")
            stat = ImageStat.Stat(rgb_img)
            # Average perceived brightness (0 to 255)
            # Standard relative luminance: 0.299 R + 0.587 G + 0.114 B
            r, g, b = stat.mean[0], stat.mean[1], stat.mean[2]
            luminosity = 0.299 * r + 0.587 * g + 0.114 * b
            metrics["mean_luminosity"] = round(luminosity, 1)

            if luminosity < 38.0:
                issues.append("Image is too dark (underexposed). Leaf lesions cannot be distinguished from shadows.")
                recommendations.append("Place one leaf in good lighting and avoid dark shadows.")
            elif luminosity > 225.0:
                issues.append("Image is overexposed (excessive sun glare). White wash obscures leaf venation.")
                recommendations.append("Shield leaf from direct blinding glare or photograph under even natural daylight.")

            # Blur detection using Laplacian variance
            gray = img.convert("L")
            if np is not None:
                arr = np.array(gray, dtype=np.float32)
                # Compute discrete 2D Laplacian operator manually if OpenCV is not installed
                # [[0, 1, 0], [1, -4, 1], [0, 1, 0]]
                if arr.shape[0] >= 3 and arr.shape[1] >= 3:
                    lap = (
                        arr[0:-2, 1:-1] +
                        arr[2:, 1:-1] +
                        arr[1:-1, 0:-2] +
                        arr[1:-1, 2:] -
                        4 * arr[1:-1, 1:-1]
                    )
                    blur_score = float(np.var(lap))
                    metrics["blur_score"] = round(blur_score, 1)

                    if blur_score < blur_threshold:
                        issues.append("Image appears blurry or out of focus. Fine fungal spots cannot be screened accurately.")
                        recommendations.append("Keep the camera steady, tap to focus on the leaf surface, and avoid motion.")
            else:
                metrics["blur_score"] = 100.0

    except Exception as e:
        return QualityCheckResult(
            is_valid=False,
            issues=[f"Corrupted image file or unreadable data: {str(e)}"],
            recommendations=["Select another valid JPEG/PNG image file."],
            metrics=metrics
        )

    # General standard instructions
    if not recommendations:
        recommendations = [
            "Place one leaf in good lighting.",
            "Avoid shadows.",
            "Keep the leaf in focus."
        ]

    is_valid = len(issues) == 0
    return QualityCheckResult(is_valid=is_valid, issues=issues, recommendations=recommendations, metrics=metrics)
